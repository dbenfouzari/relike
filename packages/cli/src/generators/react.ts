import { paramCase, snakeCase } from "change-case";
import fs from "fs";
import path from "path";

import { replace, replaceFileName } from "../utils/template";

/**
 * Generates component index file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateIndexFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateIndexFile = (key: string, componentName: string) => {
  const templateIndexPath = path.resolve(__dirname, "../templates/index/index.ts.txt");

  const templateIndex = fs.readFileSync(templateIndexPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateIndexPath, key, componentName)),
    content: replace(templateIndex, key, componentName, paramCase),
  };
};

/**
 * Generates component style file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStyleFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStyleFile = (key: string, componentName: string) => {
  const templateStylePath = path.resolve(__dirname, "../templates/react-style/__name__.module.scss.txt");

  const templateStyle = fs.readFileSync(templateStylePath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStylePath, key, componentName)),
    content: replace(templateStyle, key, componentName, snakeCase),
  };
};

/**
 * Generates component story file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateStoryFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateStoryFile = (key: string, componentName: string) => {
  const templateStoryPath = path.resolve(__dirname, "../templates/react-story/__name__.stories.tsx.txt");

  const templateStory = fs.readFileSync(templateStoryPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateStoryPath, key, componentName)),
    content: replace(templateStory, key, componentName),
  };
};

/**
 * Generates component test file
 *
 * @param   {string} key           The key to replace
 * @param   {string} componentName The component name
 * @example
 * generateTestFile("name", name);
 * @returns {any}                  File path and content.
 */
export const generateTestFile = (key: string, componentName: string) => {
  const templateTestPath = path.resolve(__dirname, "../templates/react-test/__name__.test.tsx.txt");

  const templateTest = fs.readFileSync(templateTestPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateTestPath, key, componentName)),
    content: replace(templateTest, key, componentName),
  };
};

/**
 * Generates component file
 *
 * @param   {string}  key           The key to replace
 * @param   {string}  componentName The component name
 * @param   {boolean} hasStyle      Does the component have style
 * @example
 * generateComponentFile("name", name, flags.style);
 * @returns {any}                   File path and content.
 */
export const generateComponentFile = (key: string, componentName: string, hasStyle: boolean) => {
  const templateComponentPath = hasStyle
    ? path.resolve(__dirname, "../templates/react-component/__name__(if:style).tsx.txt")
    : path.resolve(__dirname, "../templates/react-component/__name__.tsx.txt");

  const templateComponent = fs.readFileSync(templateComponentPath, { encoding: "utf-8" });

  return {
    path: path.basename(replaceFileName(templateComponentPath, key, componentName)),
    content: replace(templateComponent, key, componentName),
  };
};
