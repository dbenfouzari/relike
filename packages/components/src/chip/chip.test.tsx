import { fireEvent, render, screen } from "@testing-library/react";

import Colors from "../colors";
import Chip from "./chip";

describe("Chip", () => {
  it("should render successfully", () => {
    render(<Chip>Hello</Chip>);

    expect(screen.getByTestId("chip")).toMatchSnapshot();
  });

  it("should render successfully with dark background", () => {
    render(<Chip color={Colors.blue[900]}>Hello</Chip>);

    expect(screen.getByTestId("chip")).toMatchSnapshot();
  });

  it("should render successfully with avatar", () => {
    render(
      <Chip
        avatar={
          <img alt="avatar" src="https://wl-sympa.cf.tsp.li/resize/728x/jpg/f6e/ef6/b5b68253409b796f61f6ecd1f1.jpg" />
        }
      >
        Hello
      </Chip>,
    );

    expect(screen.getByTestId("chip")).toMatchSnapshot();
    expect(screen.getByTestId("chip-avatar")).toBeInTheDocument();
  });

  it("should handle click", () => {
    const onClick = jest.fn();

    render(<Chip onPress={onClick}>Hello</Chip>);

    fireEvent.click(screen.getByTestId("chip"));

    expect(onClick).toHaveBeenCalled();
  });

  it("should handle disabled click", () => {
    const onClick = jest.fn();

    render(
      <Chip disabled onPress={onClick}>
        Hello
      </Chip>,
    );

    fireEvent.click(screen.getByTestId("chip"));

    expect(onClick).not.toHaveBeenCalled();
  });

  it("should handle deletion", () => {
    const onDelete = jest.fn();

    render(<Chip onDeletePress={onDelete}>Hello</Chip>);

    const deleteIcon = screen.getByTestId("delete-icon");

    fireEvent.click(deleteIcon);

    expect(onDelete).toHaveBeenCalled();
  });
});
