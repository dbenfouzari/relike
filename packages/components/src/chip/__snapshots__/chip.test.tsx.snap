// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chip should render successfully 1`] = `
<button
  class="chip chip-0-2-1 chip-d0-0-2-4"
  data-testid="chip"
  tabindex="-1"
>
  <span
    class="elm-0-2-7 elm-d0-0-2-8 label"
    data-testid="text-elm"
  >
    Hello
  </span>
</button>
`;

exports[`Chip should render successfully with avatar 1`] = `
<button
  class="chip chip-0-2-1 chip-d6-0-2-13 chip__with_avatar"
  data-testid="chip"
  tabindex="-1"
>
  <div
    class="avatar"
    data-testid="chip-avatar"
  >
    <img
      alt="avatar"
      src="https://wl-sympa.cf.tsp.li/resize/728x/jpg/f6e/ef6/b5b68253409b796f61f6ecd1f1.jpg"
    />
  </div>
  <span
    class="elm-0-2-7 elm-d2-0-2-16 label"
    data-testid="text-elm"
  >
    Hello
  </span>
</button>
`;

exports[`Chip should render successfully with dark background 1`] = `
<button
  class="chip chip-0-2-1 chip-d3-0-2-9"
  data-testid="chip"
  tabindex="-1"
>
  <span
    class="elm-0-2-7 elm-d1-0-2-12 label"
    data-testid="text-elm"
  >
    Hello
  </span>
</button>
`;
